File: tmp/Square.java
import java.awt.Color;

public class Square extends Cabinet {
    private int WIDTH = 100;


    public Square() {
        super();
    }

    public Square(int x, int y) {
        super(x, y);
    }

    @Override
    public void draw() {        
        int gate = dir.ordinal();
        Main.draw.square(x, y, WIDTH-SPACING);

        int UP = (4 + (0 - gate)) % 4;
        int RIGHT = (4 + (1 - gate)) % 4;
        int DOWN = (4 + (2 - gate)) % 4;
        int LEFT = (4 + (3 - gate)) % 4;

        Main.draw.text(x, y+WIDTH-TEXT_OFFSET, (!this.isBase() || UP == 0) ? "S" + UP : ""); // UP
        Main.draw.text(x+WIDTH-TEXT_OFFSET, y, (!this.isBase() || RIGHT == 0) ? "S" + RIGHT : ""); // RIGHT
        Main.draw.text(x, y-WIDTH+TEXT_OFFSET, (!this.isBase() || DOWN == 0) ? "S" + DOWN : ""); // DOWN
        Main.draw.text(x-WIDTH+TEXT_OFFSET, y, (!this.isBase() || LEFT == 0) ? "S" + LEFT : ""); // LEFT

        if(showId) {
            Main.draw.text(x, y, String.valueOf(this.getId()));
        }
    }

    @Override
    public void highlight(Color color) {
        Main.draw.setPenColor(color);
        draw();
    }

    @Override
    public void calculatePosition(Cabinet c, int gate) {
        int pos = (c.getDir().ordinal() + gate) % 4;
        int[] vector = vectors[pos];
        int receiving_gate = adj.indexOf(c);
        int d = ((pos + 2) + (4 - receiving_gate)) % 4;
        this.dir = Direction.values()[d];
        
        this.x = c.getX() + WIDTH*2*vector[0];
        this.y = c.getY() + WIDTH*2*vector[1];
    }
}

File: app/src/main/java/com/floppa/stackcabinet/models/cabinets/Square.kt
package com.floppa.stackcabinet.models.cabinets

import androidx.compose.ui.unit.dp
import com.floppa.stackcabinet.models.Cabinet
import com.floppa.stackcabinet.models.Direction

class Square : Cabinet() {
    private val spacing = 150

    init {
        this.width = 100.dp
        this.height = 100.dp
    }

    override fun calculatePosition(cabinet: Cabinet, gate: Int) {
        val pos = (cabinet.dir?.ordinal?.plus(gate))?.rem(4)
        if (pos != null) {
            val vector: IntArray = vectors[pos]
            val receivingGate: Int = adjacent.indexOf(cabinet)
            val direction = (pos + 2 + (4 - receivingGate)) % 4
            this.dir = Direction.values()[direction]
            this.x = cabinet.x + spacing * 2 * vector[0]
            this.y = cabinet.y - spacing * 2 * vector[1]
        }
    }

    override fun toString(): String {

        val builder = StringBuilder()
        builder.append("Square: id: $id | X: $x, Y: $y | neighbors: ")
        adjacent.forEach {
            builder.append("${it?.id} ")
        }
        return builder.toString()
    }
}